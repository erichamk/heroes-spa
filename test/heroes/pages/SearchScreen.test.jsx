import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../src/heroes';
import { fireEvent, render, screen } from '@testing-library/react';
const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: () => mockUseNavigate,
}));

beforeEach(() => jest.clearAllMocks());

test('debe de mostrarse correctamente con valores por defecto', () => {
	const container = render(
		<MemoryRouter>
			<SearchPage />
		</MemoryRouter>
	);
	expect(container).toMatchSnapshot();
});

test('debe de mostrar a Batman y el input con el valor del queryString', () => {
	render(
		<MemoryRouter initialEntries={['/search?q=batman']}>
			<SearchPage />
		</MemoryRouter>
	);
	const input = screen.getByRole('textbox');
	expect(input.value).toBe('batman');

	const img = screen.getByRole('img');
	expect(img.src).toContain('/assets/heroes/dc-batman.jpg');

	const showSearchDiv = screen.getByLabelText('showSearch-div');
	expect(showSearchDiv.style.display).toBe('none');
});

test('debe mostrar un error si no se encuentra el hero (batman123)', () => {
	render(
		<MemoryRouter initialEntries={['/search?q=batman123']}>
			<SearchPage />
		</MemoryRouter>
	);

	const showErrorDiv = screen.getByLabelText('showError-div');
	expect(showErrorDiv.style.display).toBe('');
});

test('debe de llamar el navigate a la pantalla nueva', () => {
	render(
		<MemoryRouter initialEntries={['/search']}>
			<SearchPage />
		</MemoryRouter>
	);
	const searchText = 'batman';
	const input = screen.getByRole('textbox');
	//para textbox que usa useForm debe llevar name y value
	fireEvent.input(input, {
		target: { name: 'searchText', value: searchText },
	});

	const form = screen.getByRole('form');
	fireEvent.submit(form);

	expect(mockUseNavigate).toHaveBeenCalledWith(
		`?q=${searchText.toLowerCase().trim()}`
	);
});
