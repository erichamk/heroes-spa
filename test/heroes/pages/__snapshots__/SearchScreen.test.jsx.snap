// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debe de mostrarse correctamente con valores por defecto 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        Search
      </h1>
      <hr />
      <div
        class="row"
      >
        <div
          class="col-5"
        >
          <h4>
            Searching
          </h4>
          <form
            aria-label="form"
          >
            <input
              autocomplete="off"
              class="form-control"
              name="searchText"
              placeholder="Search a hero"
              type="text"
              value=""
            />
            <button
              class="btn btn-outline-primary mt-1"
            >
              Search
            </button>
          </form>
        </div>
        <div
          class="col-7"
        >
          <h4>
            Results
          </h4>
          <hr />
          <div
            aria-label="showSearch-div"
            class="alert alert-primary"
          >
            Search a hero
          </div>
          <div
            aria-label="showError-div"
            class="alert alert-danger"
            style="display: none;"
          >
            No hero with 
            <b>
              '
              '
            </b>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h1>
      Search
    </h1>
    <hr />
    <div
      class="row"
    >
      <div
        class="col-5"
      >
        <h4>
          Searching
        </h4>
        <form
          aria-label="form"
        >
          <input
            autocomplete="off"
            class="form-control"
            name="searchText"
            placeholder="Search a hero"
            type="text"
            value=""
          />
          <button
            class="btn btn-outline-primary mt-1"
          >
            Search
          </button>
        </form>
      </div>
      <div
        class="col-7"
      >
        <h4>
          Results
        </h4>
        <hr />
        <div
          aria-label="showSearch-div"
          class="alert alert-primary"
        >
          Search a hero
        </div>
        <div
          aria-label="showError-div"
          class="alert alert-danger"
          style="display: none;"
        >
          No hero with 
          <b>
            '
            '
          </b>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
